Add ability to create AWS::Scheduler::Schedule triggers and specify timezone

tie624 opened the pull request:
Adds the ability to create AWS::Scheduler::Schedule triggers as an alternative to AWS::Event::Rule, which may be preferable if hitting account-wide limits on number of AWS::Event::Rule triggers per bus (Lambdas are only able to be triggered from the default bus).

Also allows you to specify a timezone to run it on local time.

Unless they need to use inputPath or inputTransformer, developers should be able to just add the method config option and seamlessly migrate to use Scheduler if the devs find themselves up against the account quotas.

Closes: https://github.com/serverless/serverless/issues/11680

Replaces: https://github.com/serverless/serverless/pull/11707

continues from the work of @joelwalden and @KillDozerX2

tie624 commented:
@medikoo I took the existing PRs and fixed the merged conflicts and filled in any missing holes. I have also tested it and it seems to be working OK. Can you please give it a review when you get a chance?

@joelwalden and @KillDozerX2 this PR includes the work you have done. So feel free to review this as well.

medikoo commented on lib/plugins/aws/package/lib/merge-iam-templates.js:
```diff
@@ -180,6 +180,16 @@ module.exports = {
     }
 
+    // allow invoking of lambda functions for scheduler
+    policyDocumentStatements[2].Resource.push(
```
And here

medikoo commented on test/unit/lib/plugins/aws/package/compile/events/schedule.test.js:
```diff
@@ -223,6 +265,49 @@ describe('test/unit/lib/plugins/aws/package/compile/events/schedule.test.js', ()
+
+    await expect(run([events[0]])).to.be.eventually.rejectedWith(
+      ServerlessError,
+      "Configuration error at 'functions.test.events.0.schedule.method': must be equal to one of the allowed values [eventBus]\n\nLearn more about configuration validation here: http://slss.io/configuration-validation"
```
Let's validate against error`code`  and not `message` (as we do in other error handling tests)

medikoo commented on test/unit/lib/plugins/aws/package/lib/merge-iam-templates.test.js:
```diff
@@ -171,6 +172,43 @@ describe('lib/plugins/aws/package/lib/mergeIamTemplates.test.js', () => {
         );
       });
+
+      it('should allow scheduler to assume role', () => {
```
Let's not introduce such tests.

All tests should be configured with `runServerless`, we can confirm on added IAM rules in context of `schedule.test.js`

tie624 commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -2,6 +2,7 @@
 
 const _ = require('lodash');
 const ServerlessError = require('../../../../../serverless-error');
+const TIME_ZONES = require('./schedule.timezones');
```
sounds reasonable. change made.

tie624 commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -125,19 +192,33 @@ class AwsCompileScheduledEvents {
                 Input = JSON.stringify(Input);
               }
-              if (Input && typeof Input === 'string') {
+              if (Input && typeof Input === 'string' && method !== 'scheduler') {
```
In `scheduler` case the resource is not being compiled in JSON form, so there is no need to escape the quotes.

tie624 commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -199,7 +324,8 @@ class AwsCompileScheduledEvents {
               const newScheduleObject = {
-                [scheduleLogicalId]: JSON.parse(scheduleTemplate),
+                [scheduleLogicalId]:
+                  method === 'scheduler' ? scheduleTemplate : JSON.parse(scheduleTemplate),
```
The way this code is written, the `AWS::Scheduler::Schedule` resource is being compiled as an object and the `AWS::Events::Rule` resource is being compiled as JSON and later parsed into an object. Do we prefer one over another? It is confusing that it's being done differently.

medikoo replied:
@tie624 we should just work with the objects I think, still it's optional to refactor logic that applies to `AWS::Events::Rule`

Ideally if both are completely isolated from each other as:

```javascript
if (scheduler)
  // Create and attach AWS::Scheduler::Schedule resource
else 
  // Create and attach AWS::Events::Rule resource
```

tie624 replied:
They have been separated out.

tie624 commented:
> @tie624 great thanks for thaking that over, and sorry for late review.
> 
> It looks as great direction, please see my comments

@medikoo I have addressed your comments and made the appropriate changes. Can you please take another look when you can?

medikoo commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -148,69 +191,120 @@ class AwsCompileScheduledEvents {
+                  [scheduleLogicalId]: schedule,
+                };
+
+                _.merge(
```
Let's use `Object.assign` (ideally `_.merge` should be replaced in most places it's used, but ofc it's not scope of this PR)

medikoo commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -110,7 +124,16 @@ class AwsCompileScheduledEvents {
+                'Fn::GetAtt': ['IamRoleLambdaExecution', 'Arn'],
+              };
+
+              method = METHOD_EVENT_BUS;
+              if (event.schedule.method === METHOD_SCHEDULER) {
+                method = METHOD_SCHEDULER;
+              }
```
Let's simplify to:

```javascript
method  = event.schedule.method || METHOD_EVENT_BUS
```

medikoo commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -148,69 +191,120 @@ class AwsCompileScheduledEvents {
                 this.provider.naming.getLambdaSchedulePermissionLogicalId(
                   functionName,
                   scheduleNumberInFunction
                 );
```
Let's move resolution of that into below `if/else` clause

medikoo commented on lib/plugins/aws/package/lib/merge-iam-templates.js:
```diff
@@ -109,6 +109,11 @@ module.exports = {
     }-${this.provider.getStage()}`;
     const logGroupsPrefix = this.provider.naming.getLogGroupName(canonicalFunctionNamePrefix);
 
+    const assumeRolePolicyDocumentStatements =
```
Let's do these changes in the context of the event module (as e.g. we do in the case of `cloudFront` implementation -> https://github.com/serverless/serverless/blob/a6bcc62182b6fe974d12d547c00e560882c55e05/lib/plugins/aws/package/compile/events/cloud-front.js#L614-L627)

tie624 replied:
done

medikoo reviewed (changes requested):
Thank you @tie624 that looks really good!

I have just a few I believe final suggestions, and we should be good to go 🎉

tie624 commented:
> Thank you @tie624 that looks really good!
> 
> I have just a few I believe final suggestions, and we should be good to go 🎉

@medikoo thanks for the tips and direction. The changes have been made. Please take another look at it.

tie624 commented:
I also made the permissions more granular rather than using a wildcard to allow all functions.

medikoo commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -148,74 +195,170 @@ class AwsCompileScheduledEvents {
-                ${dependsOn ? `"DependsOn": "${dependsOn}",` : ''}
-                "Properties": {
-                  "FunctionName": ${lambdaTarget},
-                  "Action": "lambda:InvokeFunction",
-                  "Principal": "events.amazonaws.com",
-                  "SourceArn": { "Fn::GetAtt": ["${scheduleLogicalId}", "Arn"] }
+                if (Description) {
+                  Object.assign(properties, { Description });
+                }
+                if (timezone) {
+                  Object.assign(properties, { ScheduleExpressionTimezone: timezone });
+                }
+                if (dependsOn) {
+                  Object.assign(properties, { DependsOn: dependsOn });
                 }
-              }
-            `;
 
-              const newScheduleObject = {
-                [scheduleLogicalId]: JSON.parse(scheduleTemplate),
-              };
+                const schedule = {
+                  Type: 'AWS::Scheduler::Schedule',
+                  Properties: properties,
+                };
 
-              const newPermissionObject = {
-                [lambdaPermissionLogicalId]: JSON.parse(permissionTemplate),
-              };
+                const newScheduleObject = {
+                  [scheduleLogicalId]: schedule,
+                };
+
+                Object.assign(resources, newScheduleObject);
```
Let's simplify all of this to following notation:

```javascript
resources[scheduleLogicalId] = {
  Type: 'AWS::Scheduler::Schedule',
  Properties: {
    ScheduleExpression,
    State,
    Target: targetObj,
    FlexibleTimeWindow: { Mode: 'OFF' },
    Name,
    Description
    ...
}
```

tie624 replied:
good suggestion

medikoo commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -148,74 +195,170 @@ class AwsCompileScheduledEvents {
       }
     });
+
+    if (hasSchedulerEvents) {
```
Let's include this in already existing _if_ clause

medikoo commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -148,74 +195,170 @@ class AwsCompileScheduledEvents {
+
+    if (hasSchedulerEvents) {
+      if (!resources.IamRoleLambdaExecution) {
+        log.notice(
```
Let's make it `info` level (so it appears only in verbose mode - otherwise it feels a bit too noisy)

medikoo reviewed (changes requested):
Thanks for update @tie624. I have just a few last style suggestions

tie624 commented:
@medikoo changes made

Str77781 reviewed (commented):
# These commands leave the set-user-ID and
# set-group-ID bits of the subdirectories alone,
# so that they retain their default values.
mkdir A B C
chmod 755 A
chmod 0755 B
chmod u=rwx,go=rx C
mkdir -m 755 D
mkdir -m 0755 E
mkdir -m u=rwx,go=rx F

medikoo commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -148,69 +192,150 @@ class AwsCompileScheduledEvents {
-              let templateScheduleExpression;
-              if (typeof ScheduleExpression === 'string') {
-                templateScheduleExpression = `"${ScheduleExpression}"`;
+                const targetObj = {
+                  Arn: lambdaTarget,
+                  RoleArn: roleArn,
+                };
+                if (Input) {
+                  Object.assign(targetObj, { Input });
+                }
```
We can put it directly in below structure as:

```javascript
Target: {
  Arn: lambdaTarget,
  RoleArn: roleArn,
  Input
}
```

medikoo commented on lib/plugins/aws/package/compile/events/schedule.js:
```diff
@@ -148,69 +192,150 @@ class AwsCompileScheduledEvents {
+                      resources.IamRoleLambdaExecution.Properties.Policies[0].PolicyDocument
+                        .Statement;
+
+                    allowedInvokeResources = [];
+
+                    policyDocumentStatements.push({
+                      Effect: 'Allow',
+                      Action: ['lambda:InvokeFunction'],
+                      Resource: allowedInvokeResources,
+                    });
+                  }
+                }
+
+                if (allowedInvokeResources) {
```
Resolution of `allowedInvokeResources`, looks quite confusing, why simply not:

```javascript
if (!resources.IamRoleLambdaExecution) {
  ...
} else {
  ...
  policyDocumentStatements.push({
    Effect: 'Allow',
    Action: ['lambda:InvokeFunction'],
    Resource: [
      { 'Fn::Sub': functionArnWithVars },
      { 'Fn::Sub': `${functionArnWithVars}:*` }
    ],
  });
}
```

tie624 replied:
Actually, the context of that block only executes the first time it encounters a scheduler event. That is why pushing of resources have to happen outside of that block.

medikoo replied:
Ah indeed, forgot about that.

I believe then this needs to be handled as we do in other cases, where we add needed IAM configuration after we processed all function and events (ofc assuming that at least one event in this mode was configured)

tie624 replied:
Can you give an example of another case where we do this?

medikoo commented on test/unit/lib/plugins/aws/package/compile/events/schedule.test.js:
```diff
@@ -223,9 +272,130 @@ describe('test/unit/lib/plugins/aws/package/compile/events/schedule.test.js', ()
+    expect(scheduleCfResources[9].Properties.Target.InputTransformer).to.be.undefined;
+  });
+
+  it('should throw when passing "inputPath" or "inputTransformer" to method:schedule resources', async () => {
```
This `it` should not be in context of this `describe` (e.g. it doesn't require the `before` job as configured here).

I think to have it clean, we need to reorganize this describe:
1. Move tests that confirm on the outcome of the first `run`, into nested `describe('basic', () => ... )`  block
2. Reorganize following tests as follows:
   1. If outcome of `run` is tested in context of one `it` block, it can remain to be configured with one `it`
   2. If outcome of `run` is tested with multiple `it` blocks, let's wrap it into `describe`

tie624 replied:
I will reorganize these tests.

medikoo reviewed (changes requested):
Thank you @tie624 and sorry, there are still few thing which we should improve before merging this in

tie624 commented:
@medikoo please take another look.

medikoo reviewed (approved):
Great work @tie624 ! 🎉
